src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,1005 @@
+        private readonly LinkedList<AssignedDownload> assignedDownloadsSorted;
+
+        /// <summary>Assigned headers mapped by peer ID.</summary>
+        /// <remarks>This object has to be protected by <see cref="assignedLock"/>.</remarks>
+        private readonly Dictionary<int, List<ChainedHeader>> assignedHeadersByPeerId;
+
+        /// <summary>Block puller behaviors mapped by peer ID.</summary>
+        /// <remarks>This object has to be protected by <see cref="peerLock"/>.</remarks>
+        private readonly Dictionary<int, IBlockPullerBehavior> pullerBehaviorsByPeerId;
+
+        private readonly CancellationTokenSource cancellationSource;
+
+        /// <summary>The average block size in bytes calculated used up to <see cref="AverageBlockSizeSamplesCount"/> most recent samples.</summary>
+        /// <remarks>This object has to be protected by <see cref="queueLock" />.</remarks>
+        private readonly AverageCalculator averageBlockSizeBytes;
+
+        /// <summary>Amount of samples that should be used for average block size calculation.</summary>
+        private const int AverageBlockSizeSamplesCount = 1000;
+
+        /// <summary>The minimal count of blocks that we can ask for simultaneous download.</summary>
+        private const int MinimalCountOfBlocksBeingDownloaded = 10;
+
+        /// <summary>Signaler that triggers <see cref="reassignedJobsQueue"/> and <see cref="downloadJobsQueue"/> processing when set.</summary>
+        /// <remarks>This object has to be protected by <see cref="queueLock"/>.</remarks>
+        private readonly AsyncManualResetEvent processQueuesSignal;
+
+        /// <summary>Unique identifier which will be set to the next created download job.</summary>
+        /// <remarks>This object has to be protected by <see cref="queueLock"/>.</remarks>
+        private int nextJobId;
+
+        /// <summary>Locks access to <see cref="pullerBehaviorsByPeerId"/>.</summary>
+        private readonly object peerLock;
+
+        /// <summary>
+        /// Locks access to <see cref="processQueuesSignal"/>, <see cref="downloadJobsQueue"/>, <see cref="reassignedJobsQueue"/>,
+        /// <see cref="maxBlocksBeingDownloaded"/>, <see cref="nextJobId"/>, <see cref="averageBlockSizeBytes"/>.
+        /// </summary>
+        private readonly object queueLock;
+
+        /// <summary>Locks access to <see cref="assignedDownloadsByHash"/>, <see cref="assignedHeadersByPeerId"/>, <see cref="assignedDownloadsSorted"/>.</summary>
+        private readonly object assignedLock;
+
+        /// <summary>
+        /// The maximum blocks that can be downloaded simultaneously.
+        /// Given that all peers are on the same chain they will deliver that amount of blocks in 1 seconds.
+        /// </summary>
+        /// <remarks>This object has to be protected by <see cref="queueLock"/>.</remarks>
+        private int maxBlocksBeingDownloaded;
```

Comment:
Explain this to me, the is the count of blocks that will take to download in second based on avg calculation?

Reply:
This is the count of blocks that can be downloaded at the same time. 
It's based on avg block size and avg speed of all peers. Basically it's the amount of blocks which we expect to be able to download in 1 second.